Add pull but don't deploy, notification system

imjosh opened the pull request:
Sorry for the huge PR.  I added some features that I needed and made a few small improvements.

imjosh committed:
move config.json, update Dockerfile

put config.json in its own folder so it can be easily exposed in a bind mount

update dockerfile
- docker secret support
- take advantage of image layer caching

add dockerignore file to reduce image size

imjosh committed:
add custom logger

Add custom logger for better log formatting

index.js
-remove unused `Package` variable
-consistent quotes

imjosh committed:
add eslint

imjosh committed:
move config code to lib

index.js
-remove configuration code and move to lib/config

lib/config
-add support for reading docker secrets
-sanity check config.json

imjosh committed:
fix https://github.com/iaincollins/docker-deploy-webhook/issues/3

Add support for optional credentials

imjosh committed:
move docker api code

move docker api code out of index.js into lib/docker

imjosh committed:
promisify child_process

imjosh committed:
add pullButDontDeploy

Add options to be able to pull updated images without automatically deploying them

imjosh committed:
add stdout/err logging option

Add optional feature to log stderr and stdout from docker calls

imjosh committed:
pull before deploy

IME, docker service update doesn't reliably pull the latest the latest image.  Add workaround.

imjosh committed:
add notifications system

Added functionality to be able to send success/fail notifications via a notfication plugin system.

Added Mailgun email plugin

imjosh committed:
update README

imjosh committed:
add notify

imjosh committed:
add postinstall script

postinstall script will run NPM install against the plugins folder

imjosh committed:
remove redundant copy

from `COPY . .` the `config` directory and the `config.json` file
already exist

imjosh committed:
install `recursive-install`

Previously forgot to --save

imjosh committed:
HTTPS, pull config from Github

Add: HTTPS Server
Add: Get config.json from Github
Improve: add more error guards, logging
Fix: issues installing npm packages in plugins
Fix: don't attempt docker login if credentials weren't provided

imjosh committed:
Get mailgun config from env/files

imjosh committed:
add mailgun vars to initLocalDev

iaincollins commented:
Wow, thanks for these! 

I've been away from the keyboard for a week but will check out all the changes this week!

imjosh committed:
don't stringify Error objects

imjosh committed:
add sample compose.yml file

imjosh committed:
fix configFilePath

fix configFilePath
add CONFIG_DIR & CONFIG_FILE to initLocalDev
remove CONFIG_PATH from initLocalDev

iaincollins commented:
Hey there! I spoke to soon and turns I'm going to be out of town again until next week, will follow up Tuesday or Wednesday. Thanks! :-)

iaincollins commented:
Okay, I lied about that it seems - am going to rustle up some tests then merge this!

imjosh commented:
Are you still planning on merging?

iaincollins commented:
Hi again. Oops, I forgot about this as was looking other projects. 🤦‍♂️

I'll try and get this merged in when I get a minute to take a look.

msuzoagu committed:
adds ability to dynamically set config env variable (#2)

Why:
* current setup does not allow for dynamically setting what config to read via Dockerfile directly or via use of env variables in docker-compose file
* given multiple instances/environemts, need to be able to use same Dockerfile to set what config to read dyamically

This change address the need by:
* introducing the use of ARG values in Dockerfile to set value of CONFIG env variable at image build time

exapsy commented:
@iaincollins I think the 3000 additions probably overloaded you :P They seem like good additions for a more robust project. Could we review this?
